rust/src/socks/parser.rs

```diff
@@ -0,0 +1,160 @@
+ *
+ * You should have received a copy of the GNU General Public License
+ * version 2 along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA
+ * 02110-1301, USA.
+ */
+
+use nom7::{
+    bytes::streaming::take,
+    combinator::verify,
+    multi::count,
+    number::streaming::{be_u16, be_u8},
+    IResult,
+    error::{make_error, ErrorKind},
+    Err,
+};
+
+pub struct SocksConnectRequest {
+    pub _ver: u8,
+    pub auth_methods: Vec<u8>,
+}
+
+#[derive(Debug)]
+pub struct SocksAuthRequest<'a> {
+    pub user: &'a [u8],
+    pub pass: &'a [u8],
+}
+
+pub fn parse_connect_request(i: &[u8]) -> IResult<&[u8], SocksConnectRequest> {
+    let (i, ver) = verify(be_u8, |&v| v == 5)(i)?;
+    let (i, n) = be_u8(i)?;
+    let (i, auth_methods_vec) = count(be_u8, n as usize)(i)?;
+    let record = SocksConnectRequest {
+        _ver: ver,
+        auth_methods: auth_methods_vec,
+    };
+    Ok((i, record))
+}
+
+pub fn parse_connect_response(i: &[u8]) -> IResult<&[u8], u8> {
+    let (i, _ver) = verify(be_u8, |&v| v == 5)(i)?;
+    let (i, method) = be_u8(i)?;
+    Ok((i, method))
+}
+
+pub fn parse_auth_request(i: &[u8]) -> IResult<&[u8], SocksAuthRequest> {
+    let (i, _subver) = be_u8(i)?;
+    let (i, len) = be_u8(i)?;
+    let (i, user) = take(len)(i)?;
+    let (i, len) = be_u8(i)?;
+    let (i, pass) = take(len)(i)?;
+    let record = SocksAuthRequest { user, pass };
+    Ok((i, record))
+}
+
+pub fn parse_auth_response(i: &[u8]) -> IResult<&[u8], u8> {
+    let (i, _subver) = be_u8(i)?;
+    let (i, status) = be_u8(i)?;
+    Ok((i, status))
+}
+
+pub struct SocksConnectCommandRequest {
+    pub domain: Option<Vec<u8>>,
+    pub ipv4: Option<Vec<u8>>,
+    pub _ipv6: Option<Vec<u8>>, /// TODO
+    pub port: u16,
+}
+
+fn parse_connect_command_request_ipv4(i: &[u8]) -> IResult<&[u8], &[u8]> {
+    let (i, dst) = take(4_usize)(i)?;
+    Ok((i, dst))
+}
+
+fn parse_connect_command_request_ipv6(i: &[u8]) -> IResult<&[u8], &[u8]> {
+    let (i, dst) = take(16_usize)(i)?;
+    Ok((i, dst))
+}
+
+fn parse_connect_command_request_domain(i: &[u8]) -> IResult<&[u8], &[u8]> {
+    let (i, dlen) = be_u8(i)?; // domain
+    let (i, domain) = take(dlen)(i)?;
+    Ok((i, domain))
+}
+
+pub fn parse_connect_command_request(i: &[u8]) -> IResult<&[u8], SocksConnectCommandRequest> {
+    let (i, _ver) = verify(be_u8, |&v| v == 5)(i)?;
+    let (i, _cmd) = verify(be_u8, |&v| v == 1)(i)?;
+    let (i, _res) = verify(be_u8, |&v| v == 0)(i)?;
+    // RFC 1928 defines: 1: ipv4, 3: domain, 4: ipv6. Consider all else invalid.
+    let (i, t) = verify(be_u8, |&v| v == 1 || v == 3 || v == 4)(i)?;
+    let (i, dst) = if t == 1 {
+        parse_connect_command_request_ipv4(i)?
+    } else if t == 3 {
+        parse_connect_command_request_domain(i)?
+    } else if t == 4 {
+        parse_connect_command_request_ipv6(i)?
+    } else {
+        return Err(Err::Error(make_error(i, ErrorKind::Verify)));
```

Comment:
should be unreachable but the compiler is not smart enough to understand

Reply:
You could throw an `#[unreachable]` there, however it will panic if its every reached.

Reply:
think I prefer a debug validate bug on then, with a error handler. Would like to avoid any risk of reachable assertions.